text/0000-shared-layers.md

```diff
@@ -0,0 +1,117 @@
+
+
+## What use cases does it support?
+
+There are various use cases when you may want a common workspace that multiple buildpacks can collaborate on during the build process and this workspace needs to be cached, hence it can't be a temporary directory.
+
+In such a case a buildpack could create a "shared layer" i.e. subsequent buildpacks can modify its contents and the final state of the layer at the end of the build process is what is exported out.
+
+One example can be a CCache buildpack that sets up CCACHE_DIR environment variable (pointing to a directory in a shared layer), which a CMake buildpack may use for its build cache.
+
+Another example can be a pip cache buildpack that sets up a PIP_CACHE_DIR environment variable (pointing to a directory in a shared layer), which could be used by a [pip-tools](https://github.com/jazzband/pip-tools) buildpack, or a [pipenv](https://pipenv.pypa.io/en/latest/) buildpack to resolve the python dependency tree and generate a requirements.txt and subsequently the pip buildpack could install these requirement using the wheels cached during the dependency resolution.
+
+- What is the expected outcome?
+
+The spec and lifecycle is modified to support the above use cases.
+
+# What it is
+[what-it-is]: #what-it-is
+
+This RFC introduces a new top level folder structure called `<shared-layers>` which shall live on the same level as the `<layers>` directory.
+
+The structure would look like - 
+
+```
+<shared-layers> (configurable with `CNB_SHARED_LAYERS_DIR`)
+└── <escaped-buildpack-id> `CNB_BP_SHARED_LAYERS_DIR`
+    ├── <shared-layer>.toml
+    └── <shared-layer>
+        ├── <usable-directory>
+        └── env
+```
+
+Each buildpack will be passed an environment variable `CNB_BP_SHARED_LAYERS_DIR` with the location to the shared layers directory for a buildpack.
+
+The `<shared-layer>.toml` would look like - 
+
+```toml
+[types]
+  cache = false
+
+[metadata]
+# buildpack-specific data
+```
+
+The `env` directory would be used to set build time only variables for subsequent buildpacks. It will follow an API similar to [Buildpack ENV API](https://github.com/buildpacks/spec/blob/main/buildpack.md#provided-by-the-buildpacks)
+
+
+# How it Works
+[how-it-works]: #how-it-works
+
+All layers created in the `shared-layers` directory are build layers by default. The `cache` layer flag decides if the layer will be cached for the next run or not. This flag will follow the same semantics as [specified here.](https://github.com/buildpacks/spec/blob/main/buildpack.md#cached-layers). The `<shared-layer>.toml` for each layer will be restored with the same semantics as [specified here.](https://github.com/buildpacks/spec/blob/main/buildpack.md#cached-layers) with the `types` table being omitted during subsequent restores to avoid stale layers.
```

Comment:
So a buildpack author would have to set:

```TOML
types.cache = true
```

Reply:
I think we may need to explicitly require users to set `build = true` for each of these layers. Otherwise it will be impossible to distinguish between a stale layer and a new `cache = false` shared layer.

Reply:
yeah, mostly just asking for clarification of the ask. This makes sense to me.